libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -0,0 +1,572 @@
+    },
+    [](const auto&) {
+      TENZIR_UNREACHABLE();
+    });
+}
+
+auto get_as_located_string(const ast::expression& expr) -> located<std::string> {
+  auto arg = try_as<ast::constant>(expr);
+  TENZIR_ASSERT(arg);
+  auto loc = arg->get_location();
+  return match(
+    arg->value,
+    [loc](const std::string& s) {
+      return located{s, loc};
+    },
+    [](const auto&) -> located<std::string> {
+      TENZIR_UNREACHABLE();
+    });
+}
+
+auto strip_prefix(std::string name) -> std::string {
+  constexpr auto prefix = std::string_view{"tql2."};
+  if (name.starts_with(prefix)) {
+    return std::move(name).substr(prefix.size());
+  }
+  return name;
+}
+
+auto get_file(const boost::urls::url_view& url) -> std::string {
+  if (not url.segments().empty()) {
+    return url.segments().back();
+  }
+  if (url.host_type() == boost::urls::host_type::name) {
+    return url.host();
+  }
+  return {};
+}
+
+template <bool is_loading>
+auto create_pipeline_from_uri(std::string path,
+                              operator_factory_plugin::invocation inv,
+                              session ctx,
+                              const char* docs) -> failure_or<operator_ptr> {
+  using traits = from_to_trait<is_loading>;
+  /// We do this to make our lives easier in the code below
+  inv.args.front() = ast::constant{path, inv.args.front().get_location()};
+  const operator_factory_plugin* io_plugin = nullptr;
+  const operator_factory_plugin* compression_plugin = nullptr;
+  const operator_factory_plugin* rw_plugin = nullptr;
+  auto io_properties = typename traits::io_properties_t{};
+  auto rw_properties = typename traits::rw_properties_t{};
+  const auto pipeline_count
+    = std::ranges::count_if(inv.args, [](const ast::expression& expr) {
+        return is<ast::pipeline_expr>(expr);
+      });
+  if (pipeline_count > 1) {
+    diagnostic::error("`{}` accepts at most one pipeline",
+                      traits::operator_name)
+      .emit(ctx);
+    return failure::promise();
+  }
+  auto pipeline_argument = try_as<ast::pipeline_expr>(inv.args.back());
+  if (pipeline_count > 0) {
+    auto it = std::ranges::find_if(inv.args, [](const ast::expression& expr) {
+      return is<ast::pipeline_expr>(expr);
+    });
+    if (it != std::prev(inv.args.end())) {
+      diagnostic::error("pipeline must be the last argument")
+        .primary(it->get_location())
+        .secondary(inv.args.back().get_location())
+        .emit(ctx);
+      return failure::promise();
+    }
+  }
+  auto url = boost::urls::parse_uri_reference(path);
+  if (not url) {
+    diagnostic::error("invalid URI `{}`", path)
+      .primary(inv.args.front().get_location(), url.error().message())
+      .emit(ctx);
+    return failure::promise();
+  }
+  // determine loader based on schema
+  if (url->has_scheme()) {
+    std::tie(io_plugin, io_properties) = find_connector_given<is_loading>(
+      url->scheme(), path, inv.args.front().get_location(), docs, ctx);
```

Comment:
If this fails, we probably do not want to continue, because the downstream logic is not necessarily correct anymore.

For example: `from "htps://example.org"` fails with two errors: "unsupported scheme `htps`" and "no known format for extension `.org`". I would only expect the former.